crucible/src/Lang/Crucible/CFG/Transforms.hs

```diff
@@ -0,0 +1,730 @@
+
+         return newMap
+  where
+    outBlocks = nub (snd <$> paths)
+
+    sub renaming stmt = Fold.foldl' runRename stmt renaming
+    addNewBlock m b        = Map.insert (blockID b) b m
+    updateBlock ren m bid  = Map.update (doUpdate ren) bid m
+    doUpdate renaming blk  = Just $ mkBlock (blockID blk) (blockExtraInputs blk)
+                                      (blockStmts blk) (sub renaming <$> blockTerm blk)
+
+    runRename tgt renaming =
+      case (tgt, renaming) of
+        (Jump t, Labels from to)
+          | t == from -> Jump to
+        (Br p t1 t2, Labels from to)
+          | t1 == from -> Br p to t2
+          | t2 == from -> Br p t1 to
+        (Output ll a, LambdaLabels from to)
+          | Just Refl <- testEquality ll from -> Output to a
+        (VariantElim ctx a assgn, r@(LambdaLabels {})) ->
+          VariantElim ctx a (fmapFC (renameLabel r) assgn)
+        (MaybeBranch t a l1 l2, Labels from to)
+          | l2 == from -> MaybeBranch t a l1 to
+        (MaybeBranch t a l1 l2, LambdaLabels from to)
+          | Just Refl <- testEquality l1 from ->
+            MaybeBranch t a to l2
+        _ -> tgt 
+
+    renameLabel (LambdaLabels from to) ll
+      | Just Refl <- testEquality from ll = to
+    renameLabel _ l = l
+
+------------------------------------------------------------------------
+-- Loop Exit Merging
+
+earlyMergeLoops :: ( TraverseExt ext, Monad m, Show (CFG ext s init ret) )
+                => NonceGenerator m s
+                -> CFG ext s init ret
+                -> m (CFG ext s init ret)
+earlyMergeLoops ng cfg0 =
+  earlyMergeLoops' mempty cfg0
+  where
+    earlyMergeLoops' seen cfg
+      | (l : _) <- filter (unseen seen) (cfgLoops cfg)
+      = do cfg' <- earlyMergeLoop ng cfg l
+           earlyMergeLoops' (Set.insert (liHeader l) seen) cfg'
+      | otherwise
+      = return cfg
+
+    unseen seen li = liHeader li `Set.notMember` seen
+    
+
+earlyMergeLoop :: ( TraverseExt ext, Monad m )
+               => NonceGenerator m s
+               -> CFG ext s init ret
+               -> LoopInfo s
+               -> m (CFG ext s init ret)
+earlyMergeLoop ng cfg li =
+  case liHeader li of
+    LabelID l | not (null exits) -> 
+      do bmap'   <- funnelPaths ng bmap exits
+         let cfg' = cfg { cfgBlocks = Map.elems bmap' }
+         lowerUndefPass ng l cfg'
+    _ -> return cfg
+        
+  where
+    exits         = filterErrors (liEarlyExits li ++ liFooterIn li)
+    filterErrors  = filter (not . errorPath bmap)
+    bmap          = blockMap cfg
+
+errorPath :: Map.Map (BlockID s) (Block ext s ret) -> CFGEdge s -> Bool
+errorPath bmap (_, bid) =
+  case Map.lookup bid bmap of
+    Just (blockTerm -> Posd _ (ErrorStmt _)) -> True
+    _ -> False
+
+newtype PartialValue s tp = PartialValue { getPartial :: Reg s (MaybeType tp) }
+type ValueToPartialMap s  = MapF.MapF (Value s) (PartialValue s)
+  
+traverseCFG :: ( Monad m, TraverseExt ext )
```

Comment:
This seems like it might be generally reusable... maybe worth pulling into a more generic module?

Reply:
Does it make sense to move it into `Reg` (which seems to be the main module for mutable-register based CFGs)?

Reply:
Yeah, it might make sense there.